src/target/riscv/riscv.c

```diff
@@ -2771,12 +2772,16 @@ static int riscv_poll_hart(struct target *target, enum riscv_next_action *next_a
 				LOG_TARGET_DEBUG(target, "  triggered running");
 				target->state = TARGET_RUNNING;
 				target->debug_reason = DBG_REASON_NOTHALTED;
+				if (r->handle_became_running && r->handle_became_running(target) != ERROR_OK)
```

Comment:
These callbacks to handle the changed state look useful. I recommend to pass the `previous_riscv_state` to that callback, too. 

That's because the actions to take will likely be different also based on the original state. For instance, transition `halted --> running` is very different from `unavailable --> running`.

Reply:
Done.